scripts/e2e.sh

```diff
@@ -0,0 +1,127 @@
+#!/bin/bash
+
+# Copyright 2017 The Kubernetes Authors All rights reserved.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+KUBE_STATE_METRICS_PORT=3080
+KUBERNETES_VERSION=v1.8.0
+KUBE_STATE_METRICS_LOG_DIR=./log
+
+
+mkdir -p $KUBE_STATE_METRICS_LOG_DIR
+
+# setup a Kubernetes cluster
+curl -sLo minikube https://storage.googleapis.com/minikube/releases/latest/minikube-linux-amd64 && chmod +x minikube && sudo mv minikube /usr/local/bin/
+
+minikube version
+
+curl -sLo kubectl https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && chmod +x kubectl && sudo mv kubectl /usr/local/bin/
+
+export MINIKUBE_WANTUPDATENOTIFICATION=false
+export MINIKUBE_WANTREPORTERRORPROMPT=false
+export MINIKUBE_HOME=$HOME
+export CHANGE_MINIKUBE_NONE_USER=true
+mkdir $HOME/.kube || true
+touch $HOME/.kube/config
+
+export KUBECONFIG=$HOME/.kube/config
+sudo minikube start --vm-driver=none --kubernetes-version=$KUBERNETES_VERSION
+
+minikube update-context
+
+set +e
+
+is_kube_running="false"
+
+# this for loop waits until kubectl can access the api server that Minikube has created
+for i in {1..90}; do # timeout for 3 minutes
+   kubectl get po 1>/dev/null 2>&1
+   if [ $? -ne 1 ]; then
+      is_kube_running="true"
+      break
+   fi
+
+   echo "waiting for Kubernetes cluster up"
+   sleep 2
+done
+
+if [ $is_kube_running == "false" ]; then
+   minikube logs
+   echo "Kubernetes does not start within 3 minutes"
+   exit 1
+fi
+
+set -e
+
+kubectl version
+
+# set up kube-state-metrics manifests
+kubectl create -f ./kubernetes/kube-state-metrics-service-account.yaml
+
+kubectl create -f ./kubernetes/kube-state-metrics-cluster-role.yaml
+kubectl create -f ./kubernetes/kube-state-metrics-cluster-role-binding.yaml
+
+kubectl create -f ./kubernetes/kube-state-metrics-role-binding.yaml
+kubectl create -f ./kubernetes/kube-state-metrics-role.yaml
+
+kubectl create -f ./kubernetes/kube-state-metrics-deployment.yaml
+
+kubectl create -f ./kubernetes/kube-state-metrics-service.yaml
+
+# set up a kube-state-metrics instance
+make build
+./kube-state-metrics --in-cluster=false --kubeconfig=$KUBECONFIG --port=$KUBE_STATE_METRICS_PORT --log_dir=$KUBE_STATE_METRICS_LOG_DIR --logtostderr=false &
```

Comment:
Shouldn't we build the binary and container image before we start it as containers in the cluster? We primarily support running kube-state-metrics inside of the cluster, so that's the case we should be testing.

Reply:
@brancz I have thought about that. The main problem is that kube-state-metrics pod is deployed with `ClusterIP` service type. This is impossible for external tools to access the metrics endpoint of kube-state-metrics(Correct me if i am wrong).

By running kube-state-metrics in standalone mode, we can do this more easily. :)

BTW, as for the kube-state-metrics manifests, currently we can only check the syntax error with `kubectl create` because it will return with an non-zero exit code when something wrong happens.

The more desirable e2e test is we run at least one resource instance for each resource and we then access `/metrics` endpoint to check for the possible nil pointer deference panic. For now we only have kube-state-metrics manifests resource instances. This should be enhanced by another PR later.

Reply:
@brancz Opened issue #290 to track for adding more resource to e2e test.

Reply:
The proxy endpoint of the service resource through the apiserver allows us to access ClusterIP type services.

That looks something like this when you use it with `kubectl proxy`:

```
kubectl proxy &
curl http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kube-state-metrics:8080/metrics
```

Using kubectl proxy makes this easier as it takes care of AuthN/AuthZ.

Reply:
@brancz Awesome for pointing the usage of kubectl proxy. 👍 

The other reason that i want to use standalone mode is that we can easily get the log output of kube-state-metrics. This is good in case we want to check the log output. This can also accomplished by `kubectl logs`.

With e2e test, we need to check the newly kube-state-metrics binary. But in kube-state-metrics-deployment.yml, we use v1.1.0 kube-state-metrics image. In case to use the latest kube-state-metrics image, we need to:
* build the latest image
* get the latest image name and replace v1.1.0 in kube-state-metric-deployment.yml
* `kubectl create -f kube-state-metric-deployment.yml`

Actually i have thought about this when doing this PR. I will make the test more complicated and we should make sure that the image used in deployment is the latest build one.

Reply:
Small tip: kubectl also understands json, which is probably easier to modify in code than yaml (eg. yaml -> json -> modify with jq -> use with kubectl create)

Reply:
@brancz Comments addressed. PTAL.